README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# campfin <img src="man/figures/logo.png" align="right" width="120" />

<!-- badges: start -->

[![Lifecycle:
maturing](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://lifecycle.r-lib.org/articles/stages.html)
[![CRAN
status](https://www.r-pkg.org/badges/version/campfin)](https://CRAN.R-project.org/package=campfin)
![Downloads](https://cranlogs.r-pkg.org/badges/grand-total/campfin)
[![Codecov test
coverage](https://img.shields.io/codecov/c/github/irworkshop/campfin/master.svg)](https://app.codecov.io/gh/irworkshop/campfin?branch=master)
[![R build
status](https://github.com/irworkshop/campfin/workflows/R-CMD-check/badge.svg)](https://github.com/irworkshop/campfin/actions)
<!-- badges: end -->

The campfin package was created to facilitate the work being done on the
[The Accountability Project](https://www.publicaccountability.org/), a
tool created by [The Investigative Reporting
Workshop](https://investigativereportingworkshop.org/) in Washington,
DC. The Accountability Project curates, cleans, and indexes public data
to give journalists, researchers and others a simple way to search
across otherwise siloed records.

The data focuses on people, organizations and locations. This package
was created specifically to help with state-level campaign finance data,
although the tools included are useful in general database exploration
and normalization.

## Installation

You can install the released version of campfin from
[CRAN](https://cran.r-project.org/package=campfin) with:

``` r
install.packages("campfin")
```

The development version can be installed from
[GitHub](https://github.com/irworkshop/campfin) with:

``` r
# install.packages("remotes")
remotes::install_github("irworkshop/campfin")
```

## Normalize

The package was originally built to normalize geographic data using the
`normal_*()` functions, which take the messy self-reported geographic
data of a contributor, vendor, candidate, or committee and return
[normalized text](https://en.wikipedia.org/wiki/Text_normalization) that
is more searchable. They are largely wrappers around the
[stringr](https://github.com/tidyverse/stringr) package, and can call
other sub-functions to streamline normalization.

-   `normal_address()` takes a *street* address and reduces
    inconsistencies.
-   `normal_zip()` takes [ZIP
    Codes](https://en.wikipedia.org/wiki/ZIP_Code) and aims to return a
    valid 5-digit code.
-   `normal_state()` takes US states and returns a [2 digit
    abbreviation](https://en.wikipedia.org/wiki/List_of_U.S._state_abbreviations).
-   `normal_city()` takes cities and reduces inconsistencies.
-   `normal_phone()` consistently formats US telephone numbers.

Please see the vignette on normalization for an example of how these
functions are used to fix a wide variety of string inconsistencies and
make campaign finance data more consistent.

## Data

``` r
library(campfin)
library(tidyverse)
```

The campfin package contains a number of built in data frames and
strings used to help wrangle campaign finance data.

The `/data-raw` directory contains the code used to create the objects.

### zipcodes

The `zipcodes` (plural) table is a new version of the `zipcode`
(singular) table from the archived
[zipcode](https://cran.r-project.org/src/contrib/Archive/zipcode/) R
package.

> This database was composed using ZIP code gazetteers from the US
> Census Bureau from 1999 and 2000, augmented with additional ZIP code
> information The database is believed to contain over 98% of the ZIP
> Codes in current use in the United States. The remaining ZIP Codes
> absent from this database are entirely PO Box or Firm ZIP codes added
> in the last five years, which are no longer published by the Census
> Bureau, but in any event serve a very small minority of the population
> (probably on the order of .1% or less). Although every attempt has
> been made to filter them out, this data set may contain up to .5%
> false positives, that is, ZIP codes that do not exist or are no longer
> in use but are included due to erroneous data sources.

The included `valid_city` and `valid_zip` vectors are sorted, unique
columns from the `zipcodes` data frame.

``` r
sample_frac(zipcodes)
#> # A tibble: 44,336 × 3
#>    city       state zip  
#>    <chr>      <chr> <chr>
#>  1 SAN JUAN   PR    00914
#>  2 BRANCHDALE PA    17923
#>  3 ATHENS     IL    62613
#>  4 ALBANY     GA    31706
#>  5 HULL       IA    51239
#>  6 CHICAGO    IL    60640
#>  7 WASHINGTON DC    20380
#>  8 LA HONDA   CA    94020
#>  9 POMONA     CA    91767
#> 10 OSHKOSH    NE    69190
#> # … with 44,326 more rows
```

### `usps_*` and `valid_*`

The `usps_*` data frames were scraped from the official United States
Postal Service (USPS) [Postal Addressing
Standards](https://pe.usps.com/text/pub28/28apc_002.htm). These data
frames are designed to work with the abbreviation functionality of
`normal_address()` and `normal_city()` to replace common abbreviations
with their full equivalent.

`usps_city` is a curated subset of `usps_state`, whose full version
appear at least once in the `valid_city` vector from `zipcodes`. The
`valid_state` and `valid_name` vectors contain the columns from
`usps_state` and include territories not found in R’s build in
`state.abb` and `state.name` vectors.

``` r
sample_n(usps_street, 3)
#> # A tibble: 3 × 2
#>   full   abb  
#>   <chr>  <chr>
#> 1 PLAIN  PLN  
#> 2 COVE   CV   
#> 3 ARCADE ARC
sample_n(usps_state, 3)
#> # A tibble: 3 × 2
#>   full      abb  
#>   <chr>     <chr>
#> 1 UTAH      UT   
#> 2 ALABAMA   AL   
#> 3 WISCONSIN WI
setdiff(valid_state, state.abb)
#>  [1] "AS" "AA" "AE" "AP" "DC" "FM" "GU" "MH" "MP" "PW" "PR" "VI"
```

------------------------------------------------------------------------

The campfin project is released with a [Contributor Code of
Conduct](https://www.contributor-covenant.org/version/2/0/code_of_conduct.html).
By contributing, you agree to abide by its terms.

<!-- refs: start -->
<!-- refs: end -->


NEWS.md
# 1.0.11

* Add "PLACE" and "PL" to the `usps_street` data frame.
* Skip `non_ascii()` tests on OS other than Linux.
* Fix package help documentation.

# campfin 1.0.10

* Make `use_diary()` more flexible with custom file name argument.
* Remove the flawed character vector method for the `dplyr::count()` function.

# campfin 1.0.9

* Update template diary to match changes to TAP GitHub repository.
* Do not remove "AA" when using `normal_state(c("AA", "XX"), na_rep = TRUE)`.
* The `end` argument of `str_normal()` can now be controlled with `abb_end` in
  `normal_address()` (default `TRUE`).
* Simplify `normal_address()` by leaving number/letter mixes alone.
* Change the `punct` argument of `str_normal()` to take a replacement string.

# campfin 1.0.8

* Add `string` argument to `guess_delim()` to read the input as a single line
  of a file regardless of `\n` presence.
* Have the `delim` argument of `read_names()` default to `guess_delim()`.
* Add `pad` argument to `normal_zip()` (default `FALSE`) to control the use of
  `str_pad()` on ZIP codes without a leading zero.
* The `end` argument has been added to `abbrev_full` to target only `full`
  values at the _end_ of a string. Defaults to `FALSE`.
* `normal_address()` uses `end = TRUE` in `abbrev_full` to target only the
  street type.
  
```
abbrev_full("123 MOUNTAIN ROAD", full = usps_street, end = FALSE)
#> "123 MTN RD"
normal_address("123 MOUNTAIN ROAD", abb = usps_street)
#> "123 MOUNTAIN RD"
```

# campfin 1.0.7

* Fix testing issue with `non_ascii()` checking.
* Tweak the default template diary for `use_diary()`.

# campfin 1.0.6

* Deprecate `col_date_usa()` in favor of `col_date_mdy()`.
* `normal_address()` now only puts spaces between letters and numbers if the
  string either _starts_ with letters or _ends_ with numbers:
* `normal_address()` now keeps the forward slash in `C/O`.
* Rename `which_in()` to `what_in()` to avoid confusion with `which()`.
* Remove `http_filename()`.
* Remove `count_vec()` in favor of new `dplyr::count()` method for characters.
* Valid ZIP codes 22222, 44444, 55555 are not removed with
  `normal_zip(na.na_rep = TRUE)`.

``` r
normal_address("12east 2nd street, 3rd floor", abbs = usps_street)
#> "12 E 2ND ST 3 RD FL" # old output
#> "12 E 2ND ST 3RD FL" # fixed output
```

# campfin 1.0.4

* `file_encoding()` now fails on Solaris and tests are skipped.
    * The use of the command line `file` tool via `system2()` returns results
    on Solaris OS that are not the same as on a Unix-alike or Windows system
    and thus the results are unreliable for consistent replication.

# campfin 1.0.2

* Tests and examples run more confidently.

# campfin 1.0.0

* Improve the handling of internal data and vignette.
* Create `read_names()`.

# campfin 0.2.4

* `flag_dupes()` has an argument to flag both duplicates.
* Add encoding check to `use_diary().`
* Improve documentation examples.

# campfin 0.2.3

* `flag_dupes()` now also calls `duplicated(fromLast = TRUE)` to capture all.
* Separate address `[:digit:]` from `[:alpha:]` with space.
* Create `prop_distinct()`, `which_in()` and `which_out()`.
* Comment out examples for fetching and checking cities.
* Use 'fs' bytes and paths.
* Create `use_diary()` with template diary.
* Remove `print_all()`

# campfin 0.2.1

* The `normal_address()` now calls `abbrev_full()` instead of `expand_abbrev()`.
* Reverse order of columns in `usps_street`, `usps_state`, and `usps_city`.
* Update to version 2.0 of the `CODE_OF_CONDUCT.md` and URL.

# campfin 0.1.1

* Added a `NEWS.md` file to track changes to the package.
* Deprecated `glimpse_fun()` for more simple `col_stats()`.


R/data.R
#' @title US City, state, and ZIP
#' @description This tibble is the third version of a popular zipcodes database.
#'   The original CivicSpace US ZIP Code Database was created by Schuyler Erle
#'   using ZIP code gazetteers from the US Census Bureau from 1999 and 2000,
#'   augmented with additional ZIP code information from the Census Bureau’s
#'   TIGER/Line 2003 data set. The second version was published as the
#'   `zipcode::zipcode` dataframe object. This version has dropped the latitude
#'   and longitude, reorganized columns, and normalize the city values with
#'   [normal_city()].
#' @format A tibble with 44,336 rows of 3 variables:
#' \describe{
#'   \item{city}{Normalized city name.}
#'   \item{state}{Two letter state abbreviation.}
#'   \item{zip}{Five-digit ZIP Code.}
#'   ...
#' }
#' @source Daniel Coven's federalgovernmentzipcodes.us web site and the
#'   CivicSpace US ZIP Code Database written by Schuyler Erle
#'   <schuyler@@geocoder.us>, 5 August 2004. Original CSV files available from
#'   \url{https://web.archive.org/web/20221005220101/http://federalgovernmentzipcodes.us/free-zipcode-database-Primary.csv}
"zipcodes"

#' @title USPS Street Abbreviations
#' @description A tibble containing common street suffixes or suffix
#'   abbreviations and their full equivalent. Useful as the `add_abbs` argument
#'   of [normal_address()].
#' @format A tibble with 325 rows of 3 variables:
#' \describe{
#'   \item{full}{Primary Street Suffix.}
#'   \item{abb}{Commonly Used Street Suffix or Abbreviation.}
#'   ...
#' }
#' @source USPS Appendix
#'    [C1 Street Abbreviations](https://pe.usps.com/text/pub28/28apc_002.htm).
"usps_street"

#' @title USPS City Abbreviations
#' @description A curated and edited subset of [usps_street] containing the
#'   USPS abbreviations found in city names. Useful as the `geo_abbs` argument
#'   of [normal_city()].
#' @format A tibble with 154 rows of 2 variables:
#' \describe{
#'   \item{full}{Primary Street Suffix}
#'   \item{abb}{Commonly Used Street Suffix or Abbreviation}
#'   ...
#' }
#' @source USPS Appendix C1,
#'    [Street Abbreviations](https://pe.usps.com/text/pub28/28apc_002.htm)
"usps_city"

#' @title USPS State Abbreviations
#' @description A tibble containing the USPS.
#' @format A tibble with 62 rows of 2 variables:
#' \describe{
#'   \item{full}{Primary Street Suffix}
#'   \item{abb}{Commonly Used Street Suffix or Abbreviation}
#'   ...
#' }
#' @source USPS Appendix B,
#'    [Two–Letter State Abbreviations](https://pe.usps.com/text/pub28/28apb.htm)
"usps_state"

#' @title US State Names
#' @description The `state` column of the `usps_state` tibble.
#' @details Contains 12 more names than [datasets::state.name].
#' @format A vector of state names (length 62).
"valid_name"

#' @title US State Abbreviations
#' @description The `abb` column of the `usps_state` tibble.
#' @format A vector of 2-digit abbreviations (length 62).
"valid_state"

#' @title US State Abbreviations
#' @description The `abb` column of the `usps_state` tibble.
#' @format A vector of 2-digit abbreviations (length 62).
"valid_abb"

#' @title US City Names
#' @description The `city` column of the `zipcodes` tibble.
#' @format A sorted vector of unique city names (length 19,083).
"valid_city"

#' @title Additional US City Names
#' @description Cities not contained in [valid_city], but are
#'    accepted localities (neighborhoods or census designated
#'    places). This vector consists of normalized self-reported cities in the
#'    public data processed by accountability project that were validated
#'    by Google Maps Geocoding API (whose [check_city()] results evaluate to `TRUE`).
#'    The most recent updated version of the extra_city can be found in
#'    [this Google Sheet](https://docs.google.com/spreadsheets/d/17pi8LW1nTaGzThfUmQMZ_6HMWPxUPatqrTEWzY6LPoI)
#' @format A sorted vector of unique locality names (length 127).
"extra_city"

#' @title Invalid City Names
#' @description A custom vector containing common invalid city names.
#' @format A vector of length 54.
"invalid_city"

#' @title Almost all of the valid USA ZIP Codes
#' @description The `zip` column of the `geo` tibble.
#' @format A sorted vector of 5-digit ZIP codes (length 44334).
"valid_zip"

#' @title ZIP code regex
#' @description The regex string to extract ZIP code from the end of address.
#' @format A character string (length 1).
"rx_zip"

#' @title State regex
#' @description The regex string to extract state string preceding ZIP code.
#' @format A character string (length 1).
"rx_state"

#' @title Phone number regex
#' @description The regex string to match US phone numbers in a variety of
#'   common formats.
#' @format A character string (length 1).
"rx_phone"

#' @title URL regex
#' @description The regex string to match valid URLs.
#' @format A character string (length 1).
"rx_url"

#' @title Dark Color Palette
#' @description The Dark2 brewer color palette
#' @format A named character vector of hex color codes (length 8).
"dark2"
